### PR TITLE
Pruning Statistics Counting Fix

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
### Changes

* Logic of pruning statistic collecting was corrected (the number of pruned parameters now decreased). 
* Pruned filters statistic introduced. 
* Pruned shapes counting procedure was unified and moved to common.
* Pruning statistics was changed

### Reason for changes

Following problems had been found :
1) Input channels for depthwise convolutions wasn't removed from the layers shapes statistics (both in pruned shapes counting for statistics and for flops pruning rate choice.
2) All pruned depthwise convolutions had zero parameters with some pruned layers due to integer division on a groups number.

### Related tickets

72220

### Tests

Test cases was updated, new cases with depthwise convolution was added

### CODE REVIEW

**Comment by daniil-lyakhov** [comment]
Updated pruning model statistics
```
Statistics of the filter pruning algorithm:
+-------------------------------------------------+----------------------------+
|                Statistic's name                 |           Value            |
+=================================================+============================+
| FLOPS pruning level                             | 0.356                      |
+-------------------------------------------------+----------------------------+
| GFLOPS current / full (FLOPS current - full)    | 0.495 / 0.769 (-273745152) |
+-------------------------------------------------+----------------------------+
| MParams current / full (Params current - full)  | 2.218 / 3.956 (-1738032)   |
+-------------------------------------------------+----------------------------+
| Filters current / full (Filters current - full) | 22824 / 30348 (-7524)      |
+-------------------------------------------------+----------------------------+
| Whole model filter pruning rate                 | 0.248                      |
+-------------------------------------------------+----------------------------+
| A target level of the pruning for the algorithm | 0                          |
| for the current epoch                           |                            |
+-------------------------------------------------+----------------------------+

```

**Comment by daniil-lyakhov** [comment]
Jenkins please retry a build

**Review by vshampor — APPROVED** [review]

**Comment by alexsu52** [comment]
> Updated pruning model statistics
> 
> ```
> Statistics of the filter pruning algorithm:
> +-------------------------------------------------+----------------------------+
> |                Statistic's name                 |           Value            |
> +=================================================+============================+
> | FLOPS pruning level                             | 0.356                      |
> +-------------------------------------------------+----------------------------+
> | GFLOPS current / full (FLOPS current - full)    | 0.495 / 0.769 (-273745152) |
> +-------------------------------------------------+----------------------------+
> | MParams current / full (Params current - full)  | 2.218 / 3.956 (-1738032)   |
> +-------------------------------------------------+----------------------------+
> | Filters current / full (Filters current - full) | 22824 / 30348 (-7524)      |
> +-------------------------------------------------+----------------------------+
> | Whole model filter pruning rate                 | 0.248                      |
> +-------------------------------------------------+----------------------------+
> | A target level of the pruning for the algorithm | 0                          |
> | for the current epoch                           |                            |
> +-------------------------------------------------+----------------------------+
> ```

Some comments from my side on the new statistics table:
- What do you think about reporting (full - current) so that it has a positive value
- Preferable to report value in the same coordinates I mean if you print MParams then you should print diff (full - current) in MParams as well.
- How can explain that the target pruning level is 0, but the filter pruning rate is 0.248?
- We have 'Statistics of the pruned model' in practice the 'Pruning level of the whole model' statistic. How does 'Whole model filter pruning rate' collerate with 'Pruning level of the whole mode'?

File: nncf/common/pruning/utils.py [hunk 1]
```diff
@@ -360,6 +377,63 @@ def calculate_in_out_channels_in_uniformly_pruned_model(pruning_groups: List[Clu
...
+                                      full_output_channels,
+                                      pruning_groups_next_nodes)
+
+
+def calculate_in_out_channels_by_masks(pruning_groups: List[Cluster[PrunedLayerInfoBase]],
+                                       masks: Dict[str, NNCFTensor],
+                                       tensor_processor: NNCFBaseTensorProcessor,
+                                       full_input_channels: Dict[str, int],
+                                       full_output_channels: Dict[str, int],
+                                       pruning_groups_next_nodes: Dict[int, List[str]]):
```
Comment (alexsu52) [inline nncf/common/pruning/utils.py:413]: Please add a type hint for the return value.

**Comment by evgeniya-egupova** [comment]
@alexsu52 
> * How can explain that the target pruning level is 0, but the filter pruning rate is 0.248?

Statistics are reported before the first training epoch, so the scheduller is not updated yet and has 0 as its target. However, during initialization, the model is pruned to the `pruning_init` level.
> * We have 'Statistics of the pruned model' in practice the 'Pruning level of the whole model' statistic. How does 'Whole model filter pruning rate' collerate with 'Pruning level of the whole mode'?

`Pruning level of the whole mode` statistic from `Statistics of the pruned model` should be removed as it doesn't represent its name and is calculated differently in tf and pt.

**Comment by daniil-lyakhov** [comment]
Updated pruning statistics

```
+-------------------------------------------------+----------------------------+
|                Statistic's name                 |           Value            |
+=================================================+============================+
| FLOPS pruning level                             | 0.706                      |
+-------------------------------------------------+----------------------------+
| GFLOPS current / full (full - current)          | 0.307 / 1.044 (0.737)      |
+-------------------------------------------------+----------------------------+
| MParams current / full (full - current)         | 124.875 / 143.653 (18.778) |
+-------------------------------------------------+----------------------------+
| Filters current / full (full - current)         | 1608 / 5504 (3896)         |
+-------------------------------------------------+----------------------------+
| Filters pruning level                           | 0.292                      |
+-------------------------------------------------+----------------------------+
| A target level of the pruning for the algorithm | 0                          |
| for the current epoch                           |                            |
+-------------------------------------------------+----------------------------+
```

File: nncf/common/pruning/utils.py (hunk 1 above)
Comment (daniil-lyakhov) [reply nncf/common/pruning/utils.py]: Done

File: nncf/common/pruning/utils.py [hunk 2]
```diff
@@ -360,6 +377,63 @@ def calculate_in_out_channels_in_uniformly_pruned_model(pruning_groups: List[Clu
     return tmp_in_channels, tmp_out_channels
 
 
+def calculate_in_out_channels_in_uniformly_pruned_model(pruning_groups: List[Cluster[PrunedLayerInfoBase]],
+                                                        pruning_rate: float,
+                                                        full_input_channels: Dict[str, int],
+                                                        full_output_channels: Dict[str, int],
+                                                        pruning_groups_next_nodes: Dict[int, List[str]]):
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/common/pruning/utils.py]: Done

File: nncf/common/pruning/utils.py [hunk 3]
```diff
@@ -321,23 +330,29 @@ def count_flops_and_weights_per_node(graph: NNCFGraph,
...
+    for node in graph.get_nodes_by_metatypes(op_metatypes):
+        filters_num += output_channels.get(node.node_name, node.layer_attributes.out_channels)
+    return filters_num
+
+
+def _calculate_in_out_channels(pruning_groups: List[Cluster[PrunedLayerInfoBase]],
+                               sparse_elements_counter: Callable[[str], int],
+                               full_input_channels: Dict[str, int],
+                               full_output_channels: Dict[str, int],
+                               pruning_groups_next_nodes: Dict[int, List[str]]):
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/common/pruning/utils.py]: Done

File: nncf/common/pruning/statistics.py [hunk 4]
```diff
@@ -113,17 +117,28 @@ def __init__(self,
...
+                    f' {self.full_flops / self._giga:.3f}'
+                    f' ({self.current_flops - self.full_flops})'],
+                ['MParams current / full (Params current - full)',
+                    f'{self.current_params_num / self._mega:.3f} /'
+                    f' {self.full_params_num / self._mega:.3f}'
+                    f' ({self.current_params_num - self.full_params_num})'],
+                ['Filters current / full (Filters current - full)',
+                    f'{self.current_filters_num} / {self.full_filters_num}'
+                    f' ({self.current_filters_num - self.full_filters_num})'],
+                ['Whole model filter pruning rate', self.whole_model_pruning_level],
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/common/pruning/statistics.py]: Done

File: nncf/common/pruning/statistics.py [hunk 5]
```diff
@@ -113,17 +117,28 @@ def __init__(self,
         self.flops_pruning_level = 1 - self.current_flops / self.full_flops
         self.full_params_num = full_params_num
         self.current_params_num = current_params_num
+        self.full_filters_num = full_filters_num
+        self.current_filters_num = current_filters_num
         self.target_pruning_level = target_pruning_level
+        self.whole_model_pruning_level = 1 - self.current_filters_num/self.full_filters_num
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/common/pruning/statistics.py]: Done

**Comment by daniil-lyakhov** [comment]
Changes:
* all occurrence of names contained `pruning_rate` replaced with `pruning_level` due to both mean the same but were named different in different places. Please reach me if you disagree with such decision.
* NNCFCollectorTensorProcessor expanded by sum method and it used in generalized statistic computation 
* Refactored pruning statistics

**Comment by daniil-lyakhov** [comment]
Jenkins please retry a build

**Comment by daniil-lyakhov** [comment]
Last version of the pruning statistics:
```
Statistics of the pruned model:
+---------+---------+---------+---------------+
|    #    |  Full   | Current | Pruning level |
+=========+=========+=========+===============+
| GFLOPS  | 1.044   | 0.307   | 0.706         |
+---------+---------+---------+---------------+
| MParams | 143.653 | 124.875 | 0.131         |
+---------+---------+---------+---------------+
| Filters | 5504    | 1608    | 0.708         |
+---------+---------+---------+---------------+
Prompt: statistic pruning level = 1 - statistic current / statistic full.
Statistics of the filter pruning algorithm:
+---------------------------------------+-------+
|           Statistic's name            | Value |
+=======================================+=======+
| Filter pruning level in current epoch | 0.800 |
+---------------------------------------+-------+
| Target filter pruning level           | 0.800 |
+---------------------------------------+-------+

```

**Comment by daniil-lyakhov** [comment]
Jenkins please retry a build

**Review by AlexKoff88 — APPROVED** [review]

**Review by MaximProshin — APPROVED** [review]

**Review by AlexanderDokuchaev — APPROVED** [review]

File: nncf/common/utils/tensorboard.py [hunk 6]
```diff
@@ -43,9 +43,10 @@ def convert_to_dict(stats, algorithm_name: str):
 @convert_to_dict.register(FilterPruningStatistics)
 def _(stats, algorithm_name):
     tensorboard_stats = {
-        f'{algorithm_name}/pruning_level_for_model': stats.model_statistics.pruning_level,
-        f'{algorithm_name}/flops_pruning_level': stats.flops_pruning_level,
-        f'{algorithm_name}/target_pruning_level': stats.target_pruning_level,
+        f'{algorithm_name}/algo_current_pruning_rate': stats.current_pruning_level,
+        f'{algorithm_name}/model_FLOPS_pruning_rate': stats.model_statistics.flops_pruning_level,
+        f'{algorithm_name}/model_params_pruning_rate': stats.model_statistics.params_pruning_level,
+        f'{algorithm_name}/model_filters_pruning_rate': stats.model_statistics.filter_pruning_level,
```
Comment (daniil-lyakhov) [inline nncf/common/utils/tensorboard.py:49]:
```suggestion
        f'{algorithm_name}/algo_current_pruning_level': stats.current_pruning_level,
        f'{algorithm_name}/model_FLOPS_pruning_level': stats.model_statistics.flops_pruning_level,
        f'{algorithm_name}/model_params_pruning_level': stats.model_statistics.params_pruning_level,
        f'{algorithm_name}/model_filters_pruning_level': stats.model_statistics.filter_pruning_level,
```

File: nncf/tensorflow/pruning/callbacks.py [hunk 7]
```diff
@@ -26,9 +26,10 @@ def _prepare_for_tensorboard(self, stats: NNCFStatistics):
 
         ms = stats.filter_pruning.model_statistics
         tensorboard_stats = {
-            f'{base_prefix}/pruning_level_for_model': ms.pruning_level,
-            f'{base_prefix}/flops_pruning_level': stats.filter_pruning.flops_pruning_level,
-            f'{base_prefix}/target_pruning_level': stats.filter_pruning.target_pruning_level,
+            f'{base_prefix}/algo_current_pruning_rate': stats.filter_pruning.current_pruning_level,
+            f'{base_prefix}/model_FLOPS_pruning_rate': ms.flops_pruning_level,
+            f'{base_prefix}/model_params_pruning_rate': ms.params_pruning_level,
+            f'{base_prefix}/model_filters_pruning_rate': ms.filter_pruning_level,
```
Comment (daniil-lyakhov) [inline nncf/tensorflow/pruning/callbacks.py:32]:
```suggestion
            f'{base_prefix}/algo_current_pruning_level': stats.filter_pruning.current_pruning_level,
            f'{base_prefix}/model_FLOPS_pruning_level': ms.flops_pruning_level,
            f'{base_prefix}/model_params_pruning_level': ms.params_pruning_level,
            f'{base_prefix}/model_filters_pruning_level': ms.filter_pruning_level,
```

**Review by alexsu52 — APPROVED** [review]
Cool staff!

**Pull request merged**